### PR TITLE
Feature/LW-6653 add context to txSubmit

### PR DESCRIPTION
**Pull request by VanessaPC**
# Context
We want to continue the implementation to return and work with standard handles. 
LW-6653

# Proposed Solution
- [x] Resolve the context with handles to return HandleResolution in `SubmitTxArgs`
- [ ] ~~Replace `SignedTx` with `SubmitTxArgs`~~
- [x] add signedTx or cbor to SignedTx type
- [x] Refactor `ObservableWallet.submitTx` to take the appropriate arguments  and implement it in `PersonalWallet`
- [x] Implement the handles from context in PersonalWallet
- [x] Add tests showcasing building and submitting a transaction that sends coins 

# Important Changes Introduced
- Wallet package: 
PersonalWallet now has the handles implementation. 
ObservableWallet now takes handles. 

- Tx-construction: 
~~Replaces signedTx with SubmitTxArgs~~
Adds Cbor to SignedTx

### CODE REVIEW

File: packages/core/src/Provider/TxSubmitProvider/types.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
 // eslint-disable-next-line @typescript-eslint/ban-ts-comment
 // @ts-ignore
-import Cardano, { CardanoNodeErrors, Provider } from '../..';
+import Cardano, { CardanoNodeErrors, HandleResolution, Provider } from '../..';
 
 type SerializedTransaction = Cardano.util.HexBlob;
 
 export interface SubmitTxArgs {
   signedTransaction: SerializedTransaction;
+  context?: { handles?: HandleResolution[] | [] };
```
Comment (mkazlauskas) [inline packages/core/src/Provider/TxSubmitProvider/types.ts:9]: It's already optional, so if we we're not setting the context we can do just `{signedTransaction}` or `{signedTransaction, context: {}}`
```suggestion
  context?: { handles?: HandleResolution[] };
```

File: packages/tx-construction/src/tx-builder/types.ts [hunk 2]
```diff
@@ -139,7 +139,7 @@ export interface SignedTx {
  */
 export interface UnsignedTx {
   inspect(): Promise<TxInspection>;
-  sign(): Promise<SignedTx>;
```
Comment (mkazlauskas) [inline packages/tx-construction/src/tx-builder/types.ts:142]: Let's also delete the type that's no longer used.

File: packages/wallet/src/services/types.ts [hunk 3]
```diff
@@ -64,6 +64,7 @@ export interface OutgoingTx {
   cbor: TxCBOR;
   body: Cardano.TxBody;
   id: Cardano.TransactionId;
+  context?: HandleResolution[] | [];
```
Comment (mkazlauskas) [inline packages/wallet/src/services/types.ts:67]: Same as
```suggestion
  context?: HandleResolution[];
```

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 4]
```diff
@@ -137,19 +138,32 @@ export const DEFAULT_LOOK_AHEAD_SEARCH = 20;
...
       cbor: input,
       // Do not re-serialize transaction body to compute transaction id
       id: Cardano.TransactionId.fromTxBodyCbor(TxBodyCBOR.fromTxCBOR(input))
     };
   }
+  // SubmitTxArgs
+  if (isSubmitTxArgs(input)) {
+    return {
+      body: input.signedTransaction.body,
+      cbor: TxCBOR.serialize(input.signedTransaction),
```
Comment (mkazlauskas) [inline packages/wallet/src/PersonalWallet/PersonalWallet.ts:161]: Let's refactor `SubmitTxArgs` to `{ cbor: TxCBOR }` as we have 2 types that represent the same thing. Delete the old `SerializedTransaction` type that was used there. Please do that in a separate commit.
```suggestion
      cbor: input.signedTransaction,
```

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 5]
```diff
@@ -137,19 +138,32 @@ export const DEFAULT_LOOK_AHEAD_SEARCH = 20;
...
+      body: getSerializedTxCBOR(input),
       cbor: input,
       // Do not re-serialize transaction body to compute transaction id
       id: Cardano.TransactionId.fromTxBodyCbor(TxBodyCBOR.fromTxCBOR(input))
     };
   }
+  // SubmitTxArgs
+  if (isSubmitTxArgs(input)) {
+    return {
+      body: input.signedTransaction.body,
```
Comment (mkazlauskas) [inline packages/wallet/src/PersonalWallet/PersonalWallet.ts:160]: This shouldn't be working as `signedTransaction` should be an opaque string here :thinking: 
```suggestion
   const deserializedTx = TxCBOR.deserialize(input.signedTransaction);
   ...
      body: deserializedTx.body,
```

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 6]
```diff
@@ -137,19 +138,32 @@ export const DEFAULT_LOOK_AHEAD_SEARCH = 20;
...
+      cbor: TxCBOR.serialize(input.signedTransaction),
+      context: input?.context?.handles ?? [],
+      // Do not re-serialize transaction body to compute transaction id
+      id: input.signedTransaction.id
```
Comment (mkazlauskas) [inline packages/wallet/src/PersonalWallet/PersonalWallet.ts:164]:
```suggestion
      id: deserializedTx.id
```

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 7]
```diff
@@ -480,18 +494,20 @@ export class PersonalWallet implements ObservableWallet {
...
-    input: Cardano.Tx | TxCBOR | OutgoingTx,
+    input: Cardano.Tx | TxCBOR | OutgoingTx | SubmitTxArgs,
     { mightBeAlreadySubmitted }: SubmitTxOptions = {}
   ): Promise<Cardano.TransactionId> {
     const outgoingTx = processOutgoingTx(input);
     this.#logger.debug(`Submitting transaction ${outgoingTx.id}`);
     this.#newTransactions.submitting$.next(outgoingTx);
     try {
       await this.txSubmitProvider.submitTx({
+        ...(outgoingTx.context && { context: { handles: outgoingTx.context } }),
```
Comment (mkazlauskas) [inline packages/wallet/src/PersonalWallet/PersonalWallet.ts:505]: This reads a little weird, how about
```suggestion
        context: outgoingTx.context,
```

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 8]
```diff
@@ -480,18 +494,20 @@ export class PersonalWallet implements ObservableWallet {
...
         signedTransaction: outgoingTx.cbor
       });
       const { slot: submittedAt } = await firstValueFrom(this.tip$);
       this.#logger.debug(`Submitted transaction ${outgoingTx.id} at slot ${submittedAt}`);
+      this.#logger.debug(outgoingTx);
```
Comment (mkazlauskas) [inline packages/wallet/src/PersonalWallet/PersonalWallet.ts:510]: Let's keep a single log per submission
```suggestion
      this.#logger.debug(`Submitted transaction ${outgoingTx.id} at slot ${submittedAt}; # handles: ${outgoingTx.context?.handles?.length || 0}`);
```

File: packages/wallet/test/PersonalWallet/methods.test.ts [hunk 9]
```diff
@@ -268,6 +271,20 @@ describe('PersonalWallet methods', () => {
...
+        const txOutput = await txBuilder.buildOutput().handle('alice').coin(1_000_000n).build();
+
+        const txOut = await txBuilder.addOutput(txOutput).build().sign();
+
+        const txPending = firstValueFrom(wallet.transactions.outgoing.pending$);
+        await wallet.submitTx(txOut);
+
+        const txPendingResult = await txPending;
+
+        expect(txPendingResult.body.outputs[0].address).toEqual(mocks.resolvedHandle.resolvedAddresses.cardano);
```
Comment (mkazlauskas) [inline packages/wallet/test/PersonalWallet/methods.test.ts:286]:
```suggestion
        expect(txPendingResult.body.outputs[0].address).toEqual(mocks.resolvedHandle.resolvedAddresses.cardano);
        // expect txSubmitProvider.submitTx to be called with handle in context
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts (hunk 7 above)
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts:505]: I thought we wanted to make it optional... so if it didn't exist, we just wouldn't pass it, so we wouldn't have to modify a lot of the other transactions that consume the outgoingTx. Or that's what I understood you and Daniele agreed in the call?

Comment (mkazlauskas) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts:505]: Optional means `T | undefined`, setting it to `undefined` explicitly is fine. I understand there's a slight difference between `{}` and `{context: undefined}`, but both matches the type which is the important thing.

File: packages/tx-construction/src/tx-builder/types.ts (hunk 2 above)
Comment (VanessaPC) [reply packages/tx-construction/src/tx-builder/types.ts:142]: Is this no longer used? I think it is, but we now pass the `SubmitTxArgs` to the sign method as opposed to the `SignedArgs` to it.  So all the signedArgs types have been removed, and we now pass different ones. We could rename the method from sign to submit, but TxBuilder still uses a sign, and I think we need a sign (or submit) to get a txOut.

Comment (mkazlauskas) [reply packages/tx-construction/src/tx-builder/types.ts:142]: I think we no longer use https://github.com/input-output-hk/cardano-js-sdk/pull/751/files#diff-d2eebb6c7502396b4d246766bf4f31551ecdef4f174f4be4f9784c69bed987e3R126 type. If that's true, then there's no reason to keep it.

Comment (VanessaPC) [reply packages/tx-construction/src/tx-builder/types.ts:142]: Ah yes. I'll remove the interface.

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts (hunk 4 above)
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts:161]: While trying this refactor, I found that interestingly this breaks quite a few things. The code for the `SubmitTxArgs` has `signedTransaction` set to `SerializedTransaction` which for me is resolving to type `any` because there is no `HexBlob` type in the current import path. As a result there seem to be some places that are doing unexpected things with the `signedTransaction` assuming its actually an object. 
I think this does need to be refactored, but I think that it should be it's own ticket after having a look at it. 

Perhaps it would be better to untangle this in a separate ticket?

File: packages/wallet/test/PersonalWallet/methods.test.ts (hunk 9 above)
Comment (VanessaPC) [reply packages/wallet/test/PersonalWallet/methods.test.ts:289]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts (hunk 8 above)
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts (hunk 7 above)
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts (hunk 6 above)
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts (hunk 5 above)
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

File: packages/wallet/src/services/types.ts (hunk 3 above)
Comment (VanessaPC) [reply packages/wallet/src/services/types.ts]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

File: packages/tx-construction/src/tx-builder/types.ts (hunk 2 above)
Comment (VanessaPC) [reply packages/tx-construction/src/tx-builder/types.ts:142]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

File: packages/core/src/Provider/TxSubmitProvider/types.ts (hunk 1 above)
Comment (VanessaPC) [reply packages/core/src/Provider/TxSubmitProvider/types.ts]: 2d88f9ab41c0bf941b8126409e8d597fa2713303  should address this

**Review by mkazlauskas — APPROVED** [review]
Looking good! Please update the failing tests and squash down to a single commit for final reviews

File: packages/tx-construction/src/tx-builder/finalizeTx.ts [hunk 10]
```diff
@@ -30,7 +30,7 @@ export const finalizeTx = async (
   { ownAddresses, witness, signingOptions, auxiliaryData, isValid, handles }: TxContext,
   { inputResolver, keyAgent }: FinalizeTxDependencies,
   stubSign = false
-): Promise<SignedTx> => {
+): Promise<SubmitTxArgs> => {
```
Comment (VanessaPC) [inline packages/tx-construction/src/tx-builder/finalizeTx.ts:33]: We're changing this type to `SubmitTxArgs` but the type of `SubmitTxArgs['signedTransaction']` is meant to be an `OpaqueString`, here we're returning an object of type `Cardano.Tx`. Are we meant to be serializing the `signedTransaction` here, or is the type incorrect? @mkazlauskas

Comment (VanessaPC) [reply packages/tx-construction/src/tx-builder/finalizeTx.ts]: b15261eddb78a02ef40bd9688f6f5792babe75e9 should fix this.

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 11]
```diff
@@ -145,19 +146,33 @@ export const DEFAULT_LOOK_AHEAD_SEARCH = 20;
 // Ideally we should calculate this based on the activeSlotsCoeff and probability of a single block per epoch.
 const BLOCK_SLOT_GAP_MULTIPLIER = 2.5;
 
-const isOutgoingTx = (input: Cardano.Tx | TxCBOR | OutgoingTx): input is OutgoingTx =>
+const getSerializedTxCBOR = (input: TxCBOR) => TxCBOR.deserialize(input).body;
```
Comment (mkazlauskas) [inline packages/wallet/src/PersonalWallet/PersonalWallet.ts:149]:
```suggestion
const getSerializedTxBody = (input: TxCBOR) => TxCBOR.deserialize(input).body;
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: yarn-project.nix [hunk 12]
```diff
@@ -1327,6 +1327,7 @@ cacheEntries = {
 "fs-minipass@npm:3.0.0" = { filename = "fs-minipass-npm-3.0.0-3692c14b65-b72e9fe426.zip"; sha512 = "b72e9fe426e39f05b35bf237c8218b7ab3f68a65f325725ad7b4e431ff5a10725946fc62883b78446c07515ab938d25fdde3d08fb5ac8693f7f9eb9990da21f0"; };
 "fs.realpath@npm:1.0.0" = { filename = "fs.realpath-npm-1.0.0-c8f05d8126-99ddea01a7.zip"; sha512 = "99ddea01a7e75aa276c250a04eedeffe5662bce66c65c07164ad6264f9de18fb21be9433ead460e54cff20e31721c811f4fb5d70591799df5f85dce6d6746fd0"; };
 "fsevents@npm:2.3.2" = { filename = "fsevents-npm-2.3.2-a881d6ac9f-97ade64e75.zip"; sha512 = "97ade64e75091afee5265e6956cb72ba34db7819b4c3e94c431d4be2b19b8bb7a2d4116da417950c3425f17c8fe693d25e20212cac583ac1521ad066b77ae31f"; };
+"fsevents@patch:fsevents@npm%3A2.3.2#~builtin<compat/fsevents>::version=2.3.2&hash=18f3a7" = { filename = "fsevents-patch-3340e2eb10-8.zip"; sha512 = "edbd0fd80be379c14409605f77e52fdc78a119e17f875e8b90a220c3e5b29e54a1477c21d91fd30b957ea4866406dc3ff87b61432d2840ff8866b309e5866140"; };
```
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply yarn-project.nix:1330]: I would start with LW-6816 `Fail PR check when running `yarn` updates lockfiles`

File: packages/tx-construction/src/tx-builder/OutputBuilder.ts [hunk 13]
```diff
@@ -158,6 +153,11 @@ export class TxOutputBuilder implements OutputBuilder {
       }
     }
 
+    const outputValidation = toOutputValidationError(txOut, await this.#outputValidator.validateOutput(txOut));
```
[the remark this answers is not in the document]
Comment (VanessaPC) [reply packages/tx-construction/src/tx-builder/OutputBuilder.ts]: This was originally moved to have access to a resolved address but It's not needed now. 
883b14af431d6c40765f8350a71fa6995236fae6 should solve this

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts (hunk 11 above)
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts]: 883b14af431d6c40765f8350a71fa6995236fae6 should solve this

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 14]
```diff
@@ -145,19 +146,33 @@ export const DEFAULT_LOOK_AHEAD_SEARCH = 20;
...
-      body: TxCBOR.deserialize(input).body,
+      body: getSerializedTxCBOR(input),
       cbor: input,
       // Do not re-serialize transaction body to compute transaction id
       id: Cardano.TransactionId.fromTxBodyCbor(TxBodyCBOR.fromTxCBOR(input))
     };
   }
+  // SignedTx
+  if (isSignedTx(input)) {
+    const deserializedTx = TxCBOR.deserialize(input.cbor);
```
[the remark this answers is not in the document]
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts]: 883b14af431d6c40765f8350a71fa6995236fae6 should solve this

File: packages/wallet/src/PersonalWallet/PersonalWallet.ts [hunk 15]
```diff
@@ -145,19 +146,33 @@ export const DEFAULT_LOOK_AHEAD_SEARCH = 20;
...
+    return {
+      body: deserializedTx.body,
+      cbor: TxCBOR.serialize(input.tx),
```
[the remark this answers is not in the document]
Comment (VanessaPC) [reply packages/wallet/src/PersonalWallet/PersonalWallet.ts]: 883b14af431d6c40765f8350a71fa6995236fae6 should solve this

File: yarn-project.nix (hunk 12 above)
[the remark this answers is not in the document]
Comment (VanessaPC) [reply yarn-project.nix:1330]: This change came when I got the latest from master, when I was going to update the PR for one last round of reviews... and It doesn't go now.. I can checkout the file from master on the last rebase

**Review by mkazlauskas — DISMISSED** [review]
Nice work! Please squash for final reviews

File: yarn-project.nix (hunk 12 above)
[the remark this answers is not in the document]
Comment (mkazlauskas) [reply yarn-project.nix:1330]: I just tried running `yarn` on master and it does not flip. Must be dependent on environment (maybe os).
> https://github.com/input-output-hk/cardano-js-sdk/pull/751#discussion_r1221773728

Please do, thanks

> we should backlog this for DevX to look into

Probably needs to be fixed on nix plugin, as it's only flipping the nix lockfile (not `yarn.lock`)

File: packages/tx-construction/src/tx-builder/OutputBuilder.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (VanessaPC) [reply packages/tx-construction/src/tx-builder/OutputBuilder.ts]: @martynas when I run `yarn test:build:verify` I didn't see any error, but test failed after I pushed, so this code does need to be where it is in this line. 
However, we have a test for it, because on the tests I added we check whether the tx has a cardano.address, and if it doesn't it fails.  Which is the reason we moved it. 
I'll leave it like this, do you think it needs it further testing?

[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/tx-construction/src/tx-builder/OutputBuilder.ts:156]: If it's a `tsc` error otherwise then no need to do any further testing

[the remark this answers is not in the document]
Comment (VanessaPC) [reply packages/tx-construction/src/tx-builder/OutputBuilder.ts:156]: No, no it's not a `tsc` error, it's an actual data validation error, when there's no address, it fails.
Screenshot 2023-06-08 at 08 17 39

[the remark this answers is not in the document]
Comment (mkazlauskas) [reply packages/tx-construction/src/tx-builder/OutputBuilder.ts:156]: Ah I see, that's good too, seems to be covered by test already - nothing to do

File: yarn-project.nix (hunk 12 above)
[the remark this answers is not in the document]
Comment (VanessaPC) [reply yarn-project.nix]: d8f957d8e412600b7cdfbf0e53d5422d184bce5f <- I have just got it from master and it looks like that

**Review by mkazlauskas — APPROVED** [review]
Nice work! :rocket:

File: yarn-project.nix (hunk 12 above)
[the remark this answers is not in the document]
Comment (rhyslbw) [reply yarn-project.nix]: > Probably needs to be fixed on nix plugin, as it's only flipping the nix lockfile

Yes, but it's still in the DevX domain. I'll keep this conversation open until I've captured the work items

**Review by rhyslbw — APPROVED** [review]
Great work @VanessaPC 🚀
